### ISSUE TITLE
Implement representation for arbitrarily indexed Unions and Intersections 

### ISSUE THREAD

**Issue by hargup** [sympy/sympy#9815]
At present SymPy cannot represent arbitrarily index Unions and intersection. In symbolic notation we need ![union] and ![intersection] where ![alpha] is the indexing variable, ![I] is the index set and ![X_alpha] forms a collection of sets indexed by ![alpha].

If we allow a Set objects as input to union, then the above Union can be represented by, Intersections can be represented similarly.
```python
Union(ImageSet(Lambda(a, X(a)), I)
```

[union]: http://quicklatex.com/cache3/34/ql_32bcb1c766c59cfacf8c1ae865045534_l3.png

[intersection]: http://quicklatex.com/cache3/01/ql_7fd273d7d061055326621fd5d4d44901_l3.png

[alpha]: http://quicklatex.com/cache3/63/ql_1b9327312b7a89e41390c82ab91a2963_l3.png

[I]: http://quicklatex.com/cache3/d2/ql_6c9c97eaa68c70fa6daa0659890b4cd2_l3.png

[X_alpha]: http://quicklatex.com/cache3/f5/ql_a92029191aa0a7458c3c785473de65f5_l3.png

**Comment by jksuom** [comment]
It would be nice if we could have something like ``Union(X(a) for a in I)``, but maybe that is too hard to implement. As an alternative perhaps ``BigUnion(X(a), a, I)`` might be workable.

**Comment by hargup** [comment]
> It would be nice if we could have something like ``Union(X(a) for a in I)``, but maybe that is too hard to implement.

Agreed, it might be too hard as it will require us to overide the behaviour of `for`, done naively and we'll run into infinite loops.

**Comment by gxyd** [comment]
Is any one working on this? I would be willing to take this. I may take some time. I could not do much this week.

**Comment by jksuom** [comment]
I wasn't planning to work on the details. But I would like to hear your thoughts about having something like ``BigUnion``.

**Comment by hargup** [comment]
>  I would like to hear your thoughts about having something like ``BigUnion``.

``BigUnion`` is a pretty good name for it, thanks!


> Is any one working on this?

I was planing to work on this tonight, but it is okay you can take it. You don't have to do a lot here, just define two classes `BigUnion` and `BigIntersection`, return the ordinary union and intersection if the index set is finite and keep things unevaluated othewise. That's it. I don't think we'll be able to evaluate any operation for general case.

**Comment by Yathartha22** [comment]
@jksuom should BigUnion be implemented?

**Comment by jksuom** [comment]
I think it could be (though I have not been thinking of the details).

**Comment by Yathartha22** [comment]
Okay! I will try to look into the details.

**Comment by Yathartha22** [comment]
I have thought of the design of how `BigUnion` should work. We can implement [IndexSet](https://en.wikipedia.org/wiki/Index_set) 

```
>>> X = {FiniteSet(1, 2, 3), FinitSet(3, 4), FiniteSet(6, 7)}
I = IndexSet()
>>> BigUnion(X[i], i, I(0, 1, 2))
FiniteSet(1, 2, 3, 4, 6, 7)

>>> for i in range(0, 3):
   ....print(I[i])

Output:

FiniteSet(1, 2, 3)
FinitSet(3, 4)
FiniteSet(6, 7)

```

Same things can be applied for `ImageSets`, `Intervals`.

What do you think about this @jksuom

**Comment by jksuom** [comment]
The FiniteSet case looks plausible to me, but a union of intervals can be returned as an Interval only if it happens to be one. In general, a BigUnion class with redefined `contains` (or `_contains`) should probably be defined.

**Comment by Yathartha22** [comment]
Slight difficulty in implementing it and also visualizing the use case of Bigunion.

```python

class IndexSet(Set):

# IndexSet will return indices of the elements mapped to the set.
# Currently, the index is given implicitly.

	def __init__(self, *args):
		self.list = list(args)

	def __new__(cls, *args):
		
		if all(isinstance(i, FiniteSet) for i in args):
			indexing = {}
			for i in range(0, len(args)):
				indexing[i] = args[i]

			args = list(indexing.keys())
			return Basic.__new__(cls, *args)
		return Basic.__new__(cls, *args)


	def __getitem__(self, i):
		return self.list[i]

>>> X = IndexSet(FiniteSet(1, 2, 3), FiniteSet(3, 4), FiniteSet(6, 7)); X
IndexSet(0, 1, 2)

>>> X[0]
FiniteSet(1, 2, 3)

>>> X[1]
FiniteSet(3, 4)

>>> X[2]
FiniteSet(6, 7)
```

Now if I try something like this:

```
>>> BigUnion(X, i, FiniteSet(0, 1, 2))
{1, 2, 3, 4, 6, 7}
```

Is this the right approach? Also, how will `BigUnion` be different from `Union`, it seems what is achieved by `BigUnion` is achieved by `Union`.
